File: backend/app/models/position.py
from sqlalchemy import Column, Integer, String, Numeric, DateTime, ForeignKey, JSON, Boolean
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
import enum

from app.database import Base


class PositionStatus(str, enum.Enum):
    OPEN = "open"
    CLOSED = "closed"


class Position(Base):
    __tablename__ = "positions"

    id = Column(Integer, primary_key=True, index=True)
    ticker = Column(String(20), nullable=False, index=True)
    ticker_name = Column(String(100))
    market = Column(String(20), nullable=False)  # 'KRX', 'BINANCE'

    # Position status
    status = Column(String(20), nullable=False, default=PositionStatus.OPEN.value)
    is_info_confirmed = Column(Boolean, default=False)  # 팀장이 정보 확인/수정 완료 여부

    # Buy info
    average_buy_price = Column(Numeric(20, 4))
    total_quantity = Column(Numeric(20, 8), default=0)
    total_buy_amount = Column(Numeric(20, 2), default=0)

    # Buy plan - 분할 매수 계획 [{"price": 50000, "quantity": 10, "completed": true}, ...]
    buy_plan = Column(JSON)

    # Targets (JSONB) - completed 플래그 포함
    # [{"price": 55000, "ratio": 0.5, "completed": false}, ...]
    take_profit_targets = Column(JSON)
    stop_loss_targets = Column(JSON)

    # Sell info
    average_sell_price = Column(Numeric(20, 4))
    total_sell_amount = Column(Numeric(20, 2))

    # Performance
    profit_loss = Column(Numeric(20, 2))
    profit_rate = Column(Numeric(10, 4))
    holding_period_hours = Column(Integer)

    # Audit info
    opened_at = Column(DateTime(timezone=True))
    closed_at = Column(DateTime(timezone=True))
    opened_by = Column(Integer, ForeignKey("users.id"))
    closed_by = Column(Integer, ForeignKey("users.id"))

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    # Relationships
    opener = relationship("User", back_populates="opened_positions", foreign_keys=[opened_by])
    closer = relationship("User", back_populates="closed_positions", foreign_keys=[closed_by])
    requests = relationship("Request", back_populates="position")
    price_alerts = relationship("PriceAlert", back_populates="position")
    discussions = relationship("Discussion", back_populates="position")
    decision_notes = relationship("DecisionNote", back_populates="position", order_by="DecisionNote.created_at.desc()")

File: spring-backend/backend/src/main/java/com/fundmessenger/position/entity/Position.java
package com.fundmessenger.position.entity;

import com.fundmessenger.user.entity.User;
import io.hypersistence.utils.hibernate.type.json.JsonType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Type;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.Map;

@Entity
@Table(name = "positions")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Position {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Column(name = "ticker", length = 20, nullable = false)
    private String ticker;

    @Column(name = "ticker_name", length = 100)
    private String tickerName;

    @Column(name = "market", length = 20, nullable = false)
    private String market;

    @Column(name = "status", length = 20, nullable = false, columnDefinition = "varchar(20) default 'open'")
    private String status = "open";

    @Column(name = "is_info_confirmed", columnDefinition = "boolean default false")
    private Boolean isInfoConfirmed = false;

    @Column(name = "average_buy_price", precision = 20, scale = 4)
    private BigDecimal averageBuyPrice;

    @Column(name = "total_quantity", precision = 20, scale = 8, columnDefinition = "numeric(20,8) default 0")
    private BigDecimal totalQuantity = BigDecimal.ZERO;

    @Column(name = "total_buy_amount", precision = 20, scale = 2, columnDefinition = "numeric(20,2) default 0")
    private BigDecimal totalBuyAmount = BigDecimal.ZERO;

    @Type(JsonType.class)
    @Column(name = "buy_plan", columnDefinition = "jsonb")
    private Object buyPlan;

    @Type(JsonType.class)
    @Column(name = "take_profit_targets", columnDefinition = "jsonb")
    private Object takeProfitTargets;

    @Type(JsonType.class)
    @Column(name = "stop_loss_targets", columnDefinition = "jsonb")
    private Object stopLossTargets;

    @Column(name = "average_sell_price", precision = 20, scale = 4)
    private BigDecimal averageSellPrice;

    @Column(name = "total_sell_amount", precision = 20, scale = 2)
    private BigDecimal totalSellAmount;

    @Column(name = "profit_loss", precision = 20, scale = 2)
    private BigDecimal profitLoss;

    @Column(name = "profit_rate", precision = 10, scale = 4)
    private BigDecimal profitRate;

    @Column(name = "holding_period_hours")
    private Integer holdingPeriodHours;

    @Column(name = "realized_profit_loss", precision = 20, scale = 2, columnDefinition = "numeric(20,2) default 0")
    private BigDecimal realizedProfitLoss = BigDecimal.ZERO;

    @Column(name = "opened_at", columnDefinition = "timestamptz")
    private OffsetDateTime openedAt;

    @Column(name = "closed_at", columnDefinition = "timestamptz")
    private OffsetDateTime closedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "opened_by")
    private User opener;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "closed_by")
    private User closer;

    @CreationTimestamp
    @Column(name = "created_at", columnDefinition = "timestamptz")
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", columnDefinition = "timestamptz")
    private OffsetDateTime updatedAt;
}
